apps/crn-server/src/utils/analytics/collaboration.ts

```diff
@@ -0,0 +1,145 @@
+import { FetchUserCoproductionQuery } from '@asap-hub/contentful';
+import { TeamRole, UserCollaborationDataObject } from '@asap-hub/model';
+import { cleanArray, parseUserDisplayName } from '@asap-hub/server-common';
+
+export type EntityWithId = {
+  sys: {
+    id: string;
+  };
+};
+
+export type Author = {
+  id: string;
+  teams: EntityWithId[];
+  labs: EntityWithId[];
+};
+
+export const checkDifferentTeams = (
+  referenceTeam: EntityWithId,
+  testTeams: EntityWithId[],
+): boolean =>
+  !testTeams.find((testTeam) => testTeam.sys.id === referenceTeam.sys.id);
+
+export const checkSameTeamDifferentLab = (
+  referenceTeam: EntityWithId,
+  referenceLabs: EntityWithId[],
+  testUser: Author,
+): boolean => {
+  const sameTeam = testUser.teams.find(
+    (testTeam) => testTeam.sys.id === referenceTeam.sys.id,
+  );
+  if (sameTeam) {
+    return referenceLabs
+      .map((lab) =>
+        testUser.labs.find((testLab) => testLab.sys.id === lab.sys.id),
+      )
+      .every((value) => !value);
+  }
+  return false;
+};
+
+export const findMatchingAuthors = ({
+  referenceId,
+  referenceTeam,
+  referenceLabs,
+  authorList,
+}: {
+  referenceId: string;
+  referenceTeam: EntityWithId;
+  referenceLabs: EntityWithId[];
+  authorList: Author[];
+}) => {
+  let [differentTeamFlag, sameTeamDifferentLabFlag] = [false, false];
+  authorList.forEach((author) => {
+    if (author && author.id !== referenceId) {
+      if (
+        !differentTeamFlag &&
+        checkDifferentTeams(referenceTeam, author.teams)
+      ) {
+        differentTeamFlag = true;
+      }
+      if (
+        !sameTeamDifferentLabFlag &&
+        checkSameTeamDifferentLab(referenceTeam, referenceLabs, author)
+      ) {
+        sameTeamDifferentLabFlag = true;
+      }
+    }
+  });
+  return { differentTeamFlag, sameTeamDifferentLabFlag };
+};
+export const getCollaborationCounts = (
+  data: {
+    differentTeamFlag: boolean;
+    sameTeamDifferentLabFlag: boolean;
+  }[],
+) =>
+  data.reduce(
+    (
+      { acrossTeamCount, withinTeamCount },
+      { differentTeamFlag, sameTeamDifferentLabFlag },
+    ) => ({
+      acrossTeamCount: differentTeamFlag
+        ? acrossTeamCount + 1
+        : acrossTeamCount,
+      withinTeamCount: sameTeamDifferentLabFlag
+        ? withinTeamCount + 1
+        : withinTeamCount,
+    }),
+    {
+      acrossTeamCount: 0,
+      withinTeamCount: 0,
+    },
+  );
+
+export const getUserCoproductionItems = (
+  userCollection: FetchUserCoproductionQuery['usersCollection'],
+): UserCollaborationDataObject[] =>
+  cleanArray(userCollection?.items).map((user) => {
+    const teams = cleanArray(user?.teamsCollection?.items).map((team) => {
+      const analyticData =
+        user.linkedFrom?.researchOutputsCollection?.items.map((output) => {
+          const authorList = cleanArray(
+            cleanArray(output?.authorsCollection?.items).map((author) => {
+              if (author.__typename === 'Users') {
```

Comment:
Not all research outputs linked to a user is linked because they are authors 😅 

This one for example, returns a total of 84 outputs 

![Screenshot 2024-05-02 at 11 56 59](https://github.com/yldio/asap-hub/assets/16595804/e6d1aad1-8fd7-459c-b08b-992af3476763)

But in the hub https://hub.asap.science/network/users/1fhU4Tx6CedIzQX0js9ajD/research we see that there's 0 outputs

![Screenshot 2024-05-02 at 12 00 18](https://github.com/yldio/asap-hub/assets/16595804/790bea8e-46e2-4e91-83e1-7aebe4bedfa9)

she's linked to the outputs because she might be the one who updated it not necessarily the author. So it's necessary to check if the author is the user we are analyzing